.github/workflows/build.yml

```diff
@@ -0,0 +1,266 @@
+---
+name: build
+'on': [push, pull_request]
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+jobs:
+  linux:
+    name: linux
+    runs-on: ubuntu-latest
+    container: lmmsci/linux.gcc:18.04
+    env:
+      CMAKE_OPTS: >-
+        -DUSE_WERROR=ON
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo
+        -DUSE_COMPILE_CACHE=ON
+      CCACHE_MAXSIZE: 500M
+      MAKEFLAGS: -j2
+    steps:
+      - name: Update and configure Git
+        run: |
+          add-apt-repository ppa:git-core/ppa
+          apt-get update
+          apt-get --yes install git
+          git config --global --add safe.directory /__w/lmms/lmms
+      - name: Check out
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: recursive
+      - name: Cache ccache data
+        uses: actions/cache@v3
+        with:
+          key: ccache-${{ github.job }}-${{ github.ref }}-${{ github.run_id }}
+          restore-keys: |
+            ccache-${{ github.job }}-${{ github.ref }}-
+            ccache-${{ github.job }}-
+          path: ~/.ccache
+      - name: Configure
+        run: |
+          source /opt/qt5*/bin/qt5*-env.sh || true
+          mkdir build && cd build
+          cmake .. $CMAKE_OPTS -DCMAKE_INSTALL_PREFIX=./install
+      - name: Build
+        run: cmake --build build
+      - name: Build tests
+        run: cmake --build build --target tests
+      - name: Run tests
+        run: build/tests/tests
+      - name: Package
+        run: |
+          cmake --build build --target install
+          cmake --build build --target appimage \
+            || (cat build/appimage.log && false)
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: linux
+          path: build/lmms-*.AppImage
+      - name: Print ccache statistics
+        run: |
+          echo "[ccache config]"
+          ccache -p
+          echo "[ccache stats]"
+          ccache -s
+  macos:
+    name: macos
+    runs-on: macos-11
+    env:
+      CMAKE_OPTS: >-
+        -DUSE_WERROR=ON
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo
+        -DUSE_COMPILE_CACHE=ON
+      CCACHE_MAXSIZE: 500M
+      MAKEFLAGS: -j3
+      DEVELOPER_DIR: /Applications/Xcode_11.7.app/Contents/Developer
+    steps:
+      - name: Check out
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: recursive
+      - name: Cache ccache data
+        uses: actions/cache@v3
+        with:
+          key: ccache-${{ github.job }}-${{ github.ref }}-${{ github.run_id }}
+          restore-keys: |
+            ccache-${{ github.job }}-${{ github.ref }}-
+            ccache-${{ github.job }}-
+          path: ~/Library/Caches/ccache
+      - name: Install dependencies
```

Comment:
Maybe worth caching `~/Library/Caches/Homebrew`(package download directory) as well, like we do in CircleCI builds.

Reply:
Initially I copied the CircleCI config, which caches `~/Library/Caches/Homebrew` and `/usr/local/Homebrew`, but it turned out that GitHub's macOS image has a bunch of stuff preinstalled in Homebrew, which made the cache use about 4GB of our 10GB quota.

If this is just the download directory, I guess it should only contain what we install ourselves, right? I wonder what proportion of the installation time is spent downloading, and how much faster GitHub's cache is. Also, will this grow in size over time as newer package versions are released? Sorry, I don't have much experience with Homebrew.

Reply:
The cache is still 3.3GB with just that one directory. I'm reverting it again.